<a target="_blank" href="https://colab.research.google.com/github/BenjaminHerrera/MAT421/blob/main/Module_A.ipynb">
  <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
</a>

# **MODULE A:** Representation of Numbers
# **AUTHOR:** Benjamin Joseph L. Herrera
# **CLASS:** MAT 421
# **DATE:** 14 JAN 2024

## Representation of Numbers, Base-N, and Binary

There are many ways to represent numbers. For example, we can represent `10` via the following formats

In [28]:
# Different values!
example_1 = 10
example_2 = bin(10)
example_3 = "10"
example_4 = 0x000A

# These are all printed to the value of 10
print(example_1)
print(int(example_2, 2))
print(example_3)
print(example_4)

10
10
10
10


Decimal values are represented in base ten. With decimal values, we can decompose 
a number with powers of the base it's representing. For example, `423.23` can be 
represented below using coefficients and powers of ten (base ten):


$$423.23 = 4 \times 10^2 + 2 \times 10^1 + 3 \times 10^0 + 2 \times 10^{-1} + 3 \times 10^{-2}$$

Here's another example of this decomposition with Python:

_NOTE:_ this will result into `423.22999999999996`, so a `round()` function is applied for appearances. We'll circle to this later.

In [29]:
print(round(4 * 10**2 + 2 * 10**1 + 3 * 10**0 + 2 * 10**-1 + 3 * 10**-2, 2))

423.23


Binary values are another form of representing numbers. This is done with base two representation. Using the number `12`, we can represent it as follows:

$$12 = 1 \times 10^1 + 2 \times 10^0$$

With binary (base two), it is calculated as follows: 

$$12 = 1 \times 2^3 + 1 \times 2^2 + 0 \times 2^1 + 0 \times 2^0$$

This is binary is also represented as `1100`

When adding and multiplying two binary values, it is the same process and adding and multplying decimal values. Here's are examples of adding 11 and 13 together in binary:

$$\begin{align*}
  & \texttt{ 111\;} \\
  & \texttt{ 1011} \quad \text{(11)}\\
+ & \texttt{ 1101} \quad \text{(13)}\\
\hline
  & \texttt{11000} \quad \text{(24)}
\end{align*}$$

In [30]:
a = 0b1011
b = 0b1101
c = a + b
print("Result of ADDING 1011 + 1101 (11 + 13)")
print("Decimal: ", c)
print("Binary: ", bin(c))

Result of ADDING 1011 + 1101 (11 + 13)
Decimal:  24
Binary:  0b11000


Here are and examples of multiplying 11 and 13 together:

$$\begin{align*}
  & \texttt{~~~~1011} \quad \text{(11)} \\
\times & \texttt{~~~~1101} \quad \text{(13)} \\
\hline
  & \texttt{~~~~1011} \\
  & \texttt{~~~~~~~0} \\
  & \texttt{~~101100} \\
+ & \texttt{~1011000} \\
\hline
  & \texttt{10001111} \quad \text{(143)}
\end{align*}$$

In [31]:
a = 0b1011
b = 0b1101
c = a * b
print("Result of MULTIPLYING 1011 and 1101 (11 * 13)")
print("Decimal: ", c)
print("Binary: ", bin(c))

Result of MULTIPLYING 1011 and 1101 (11 * 13)
Decimal:  143
Binary:  0b10001111


## Floating Point Numbers

64-bit decimal values are represented using the `IEEE 754` standard. This standard uses 64 bits to represent a float value. These bits are decomposed to the following sections, in order: 
- **1 bit** for indicating the sign of the value ($s$)
- **11 bits** for the exponent portion ($e$)
- **52 bits** for the fraction portion ($f$)

To calculate the value of a float, we use the following equation:

$$n = (-1)^{s} \times 2^{e-1023} \times (1 + f)$$

Let's take an example value of `23`. How could we represent this in the IEEE 754 format? 

`23` is a positive number, so $s$ is set to 0. 

The power of two that is lower than the value of `23` is 16. Thus, $e-1023$ needs to be 4, which makes $e = 1027$.

To find the value of $f$, we solve for $f$. This leaves us with the value of `0.4375`.

We can represent $s$ as `1` and $e$ as `10000000011`. To reppresent $f$, we use the following formula:

$$f = f_{1} \cdot \frac{1}{2^1} + f_{2} \cdot \frac{1}{2^2} +  f_{3} \cdot \frac{1}{2^3} +\cdots f_{52} \cdot \frac{1}{2^{52}}$$

Thus, $f$ is `0111000000000000000000000000000000000000000000000000`.

This means that IEEE 754 representation for `23` is `1 10000000011 0111000000000000000000000000000000000000000000000000`

To go from IEEE 754 format to decimal, we simply plug the values of the three sections into $n$ from above. 

For example, we have the float representation, `0 10000000100 1010000000000000000000000000000000000000000000000000`

We can see that the $s$ is just `0` and that $e$ is `1028`. For the $f$ value, we use the $f$ formula to figure out its value:

$$f = 1 \cdot \frac{1}{2^1} + 0 \cdot \frac{1}{2^2} + 1 \cdot \frac{1}{2^3} = 0.625$$

Using $n$, we get a value of `52`.

## Round-Off Errors

The IEEE 754 standard isn't precise. We have encountered this issue when we were decomposing `423.23` into coefficients and powers of two. 

In [32]:
print(4 * 10**2 + 2 * 10**1 + 3 * 10**0 + 2 * 10**-1 + 3 * 10**-2)

423.22999999999996


Shown above is an imprecision of this system where it doesn't exactly meassure to `423.23`.  

To solve this issue, we round off the value to some set number of bytes. However, when the rounded value is not equal to its ground-truth value, we get what's called a `round-off error`.

Here's another example of round-off error:

In [33]:
3.66666666666 - 1.33333333333

2.3333333333299997

In [34]:
3.66666666666 - 1.33333333333 == 2.33333333333

False

Above is an round-off error where the subtraction operation should've resulted in the value of `1.33333333333`. Instead, it resulted in `2.3333333333299997`. 

Round-off errors can accumulate. Here's an example of this:

In [35]:
a = 0
for i in range(9):
    a += 1.33333333333
a

11.999999999969999

Here, you can see that there is a random `6` digit in the resulting `a`. This should've been a `9`, which would make the ground-truth value of `a = 11.99999999999`.